moveit_ros/moveit_servo/src/servo_calcs.cpp

```diff
@@ -758,58 +758,25 @@ double ServoCalcs::velocityScalingFactorForSingularity(const Eigen::VectorXd& co
 
 void ServoCalcs::enforceVelLimits(Eigen::ArrayXd& delta_theta)
 {
+  // Convert to joint angle velocities for checking and applying joint specific velocity limits.
   Eigen::ArrayXd velocity = delta_theta / parameters_.publish_period;
 
   std::size_t joint_delta_index = 0;
-  // Track the smallest velocity scaling factor required, across all joints
-  double velocity_limit_scaling_factor = 1;
-
-  for (auto joint : joint_model_group_->getActiveJointModels())
+  double velocity_scaling_factor = 1;
+  for (const moveit::core::JointModel* joint : joint_model_group_->getActiveJointModels())
   {
-    // Some joints do not have bounds defined
-    const auto bounds = joint->getVariableBounds(joint->getName());
-
+    const auto& bounds = joint->getVariableBounds(joint->getName());
     if (bounds.velocity_bounded_)
     {
-      velocity(joint_delta_index) = delta_theta(joint_delta_index) / parameters_.publish_period;
-
-      bool clip_velocity = false;
-      double velocity_limit = 0.0;
-      if (velocity(joint_delta_index) < bounds.min_velocity_)
-      {
-        clip_velocity = true;
-        velocity_limit = bounds.min_velocity_;
-      }
-      else if (velocity(joint_delta_index) > bounds.max_velocity_)
-      {
-        clip_velocity = true;
-        velocity_limit = bounds.max_velocity_;
-      }
-
-      // Apply velocity bounds
-      if (clip_velocity)
-      {
-        const double scaling_factor =
-            fabs(velocity_limit * parameters_.publish_period) / fabs(delta_theta(joint_delta_index));
-
-        // Store the scaling factor if it's the smallest yet
-        if (scaling_factor < velocity_limit_scaling_factor)
-          velocity_limit_scaling_factor = scaling_factor;
-      }
+      const auto unbounded_velocity = velocity(joint_delta_index);
+      velocity(joint_delta_index) = std::min(std::max(unbounded_velocity, bounds.min_velocity_), bounds.max_velocity_);
+      velocity_scaling_factor = std::min(velocity_scaling_factor, unbounded_velocity / velocity(joint_delta_index));
+      ++joint_delta_index;
```

Comment:
it seems like there is potential here (`if (!bounds.velocity_bounded_)`) that joint_delta_index doesn't get incremented properly

Reply:
should it be moved outside of the if()?

Reply:
Good catch. Would have definitely been a bug and I'll need to fix it. But this comes back to my question in the description about how are the joint bounds and `delta_theta`s guaranteed to match? Is it simply because the `joint_model_group_` is initialized once and all its class methods access the joint in the same order? I marked this as WIP because I want to convince myself that the indices and deltas do indeed match but I haven't yet had the time.